File: backend/src/echomind/schemas/rag.py
"""Schemi HTTP per il RAG (M8) -- boundary tra API e service layer.

Separati da `rag/schema.py` (tipi di dominio interni) per due motivi:
  1. `AskRequest`/`AskResponse` sono schemi HTTP (Pydantic v2, serializzazione
     JSON, validazione input): non devono essere importati dal service layer.
  2. `CitationRead` e' una proiezione pubblica di `Citation` (rag/schema.py):
     consente di aggiungere campi futuri (es. excerpt) senza toccare lo
     structured output LLM.

Vincoli:
  - `question` ha min_length=1 e max_length corrispondente a `rag_max_question_chars`
    (default 2000). La costante qui rispecchia il default di config: se si modifica
    il limite, aggiornare entrambi.
  - `citations` e' una lista vuota di default: la risposta rimane valida anche
    quando il contesto non contiene entita' rilevanti.
"""

from __future__ import annotations

from pydantic import BaseModel, Field

# Limite massimo caratteri della domanda: deve combaciare con il default di
# Settings.rag_max_question_chars in core/config.py.
_MAX_QUESTION_CHARS: int = 2000


class AskRequest(BaseModel):
    """Corpo della POST /documents/{id}/ask."""

    question: str = Field(
        min_length=1,
        max_length=_MAX_QUESTION_CHARS,
        description=(
            f"Domanda sul contenuto del documento (1-{_MAX_QUESTION_CHARS} caratteri). "
            "La risposta viene generata nella lingua indicata dall'header Accept-Language."
        ),
    )


class CitationRead(BaseModel):
    """Entita' del grafo citata come fonte nella risposta RAG."""

    entity_id: str = Field(description="ID UUID (stringa) dell'entita' citata.")
    name: str = Field(description="Nome dell'entita' nel grafo.")
    type: str = Field(description="Tipo dell'entita' (es. PERSON, LOCATION, CONCEPT).")


class AskResponse(BaseModel):
    """Risposta dell'endpoint /ask: testo in linguaggio naturale + citazioni."""

    answer: str = Field(description="Risposta completa alla domanda, nella lingua richiesta.")
    citations: list[CitationRead] = Field(
        default_factory=list,
        description=(
            "Entita' del grafo usate come fonte. Vuota se il contesto non contiene "
            "informazioni sufficienti o se la domanda non richiede citazioni specifiche."
        ),
    )

File: frontend/src/features/graph/schemas/rag.ts
/**
 * Schema Zod per la GraphRAG (Q&A per-documento).
 * Mirror di `AskResponse` / `CitationRead` in backend/rag/gemini.py.
 *
 * `Citation`:
 *   Entità del grafo usata come fonte per la risposta generata da Gemini.
 *   `entity_id` è l'UUID dell'entità in Neo4j (usato per evidenziare il nodo
 *   nel GraphViewer quando l'utente legge la risposta RAG).
 *
 * `AskResponse`:
 *   `answer`: testo in linguaggio naturale generato da Gemini con `response_schema`
 *             strutturato (vedi backend/rag/gemini.py: `GeminiRagAnswerer`).
 *   `citations`: lista di entità citate, con `.default([])` per robustezza
 *               (se il backend non le include, non si rompe la validazione Zod).
 *
 * La lingua della risposta dipende dall'header Accept-Language della richiesta
 * (impostato dall'interceptor axios con `i18n.language`).
 */
import { z } from "zod";

/** Entità del grafo citata come fonte nella risposta. */
export const citationSchema = z.object({
  entity_id: z.string(),
  name: z.string(),
  type: z.string(),
});
export type Citation = z.infer<typeof citationSchema>;

/** Risposta dell'endpoint POST /documents/{id}/ask (GraphRAG). */
export const askResponseSchema = z.object({
  answer: z.string(),
  // `.default([])` rende il campo opzionale in entrata: Zod usa [] se mancante.
  citations: z.array(citationSchema).default([]),
});
export type AskResponse = z.infer<typeof askResponseSchema>;
